### ISSUE TITLE
"func" is not a known attribute of "None" for variable dictionary classes after null check

### ISSUE THREAD

**Issue by ChiNoel-osu** [microsoft/pyright#11205]
## Environment data

-   Pylance version: 2025.10.4
-   OS and version: Windows 10 22H2
-   Python version (& distribution if applicable, e.g. Anaconda): 3.13.5

## Code Snippet

```python
class SomeClass:
    def __init__(self) -> None:
        self.nullable_str: str | None = None


d: dict[str, SomeClass] = {}

class_instance = SomeClass()
class_instance.nullable_str = "abc"

key = "key"

d[key] = class_instance

d[key].nullable_str.capitalize()

if d[key].nullable_str:
    d[key].nullable_str.capitalize()

if d[key].nullable_str is not None:
    d[key].nullable_str.capitalize()
```

## Repro Steps

1. Copy & paste code
2. Observe

## Expected behavior

Pylance reports no error.

## Actual behavior

Pylance reports `"capitalize" is not a known attribute of "None"` for all `capitalize()` calls.

<img width="1025" height="540" alt="Image" src="https://github.com/user-attachments/assets/a1e32b92-97af-4328-bcc5-4ead52490cc6" />

If replace the variable `key` to a fixed one (d[key] -> d["something"]), all errors goes away.

Please correct me if I'm just using Python wrong.

## Logs

```
Log looks fine, there's no abnormality.
```

**Comment by ChiNoel-osu** [comment]
The workaround works, but my code would be more complex just to make Pylance happy 😐.

**Comment by erictraut** [comment]
Pyright is working as designed here.

Type narrowing is not supported for dynamic (variable) subscript values. This same limitation exists for all Python type checkers. For more details about type narrowing, refer to [this documentation](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-narrowing).
